### PR TITLE
feat: add jsonrpc gateway support

### PR DESCRIPTION
**Pull request by alidevjimmy**
## Description

Added JSON-RPC gateway support for gRPC server using  [protoc-gen-jrpc-gateway](https://github.com/PacViewer/jrpc-gateway/tree/main/protoc-gen-jrpc-gateway) protoc generator plugin and [jrpc](https://github.com/PacViewer/jrpc-gateway) for json-rpc server.
I tested it manually and it works fine but I'll write tests for it.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -24,18 +24,25 @@ devtools:
...
-	go build -o ./build/pactus-daemon$(EXE) ./cmd/daemon
-	go build -o ./build/pactus-wallet$(EXE) ./cmd/wallet
-	go build -o ./build/pactus-shell$(EXE) 	./cmd/shell
+build: build_daemon build_wallet build_shell
 
 build_race:
 	go build -race -o ./build/pactus-daemon$(EXE) ./cmd/daemon
 	go build -race -o ./build/pactus-wallet$(EXE) ./cmd/wallet
 
+build_daemon:
```
Comment (b00f) [inline Makefile:37]: Explain why did you create new target per binary?

File: config/config.go [hunk 2]
```diff
@@ -119,6 +122,8 @@ func DefaultConfigMainnet() *Config {
 	conf.GRPC.BasicAuthCredential = ""
 	conf.GRPC.Gateway.Enable = false
 	conf.GRPC.Gateway.Listen = "127.0.0.1:8080"
+	conf.JSONRPC.Enable = false
```
Comment (b00f) [inline config/config.go:125]: Please update the `example_config.toml`

File: www/grpc/buf/buf.gen.yaml [hunk 3]
```diff
@@ -32,6 +32,9 @@ plugins:
     opt:
       - grpc_api_configuration=grpc-gateway.config.yaml,allow_merge=true,merge_file_name=pactus
       - openapi_configuration=openapi.config.yaml
+  - name: jrpc-gateway
+    out: ../gen/go
+    opt: paths=source_relative
```
Comment (b00f) [inline www/grpc/buf/buf.gen.yaml:37]: Is this option required?

File: www/jsonrpc/server.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type Server struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	config     *Config
+	server     *http.Server
+	grpcClient *grpc.ClientConn
+	handlers   map[string]func(ctx context.Context, params json.RawMessage) (interface{}, error)
```
Comment (b00f) [inline www/jsonrpc/server.go:23]: I think `handlers` should be remove from here and moves to plugin package.

**Review by b00f — CHANGES_REQUESTED** [review]

**Comment by b00f** [comment]
@alidevjimmy, you can use commands like `make fmt` and `make check` to ensure that linting and formatting pass.

File: www/jsonrpc/server.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			case <-s.ctx.Done():
+				return
+			default:
+				if err := server.ListenAndServe(); err != nil {
+					s.logger.Error("error while establishing JSON-RPC connection", "error", err)
+				}
+			}
+		}
+	}()
+	s.logger.Info("jsonrpc started listening", "address", s.config.Listen)
```
Comment (b00f) [inline www/jsonrpc/server.go:89]: It is better to create a listener and then serve the HTTP (line 83). 
Then you can get the exact listening address from the listener, especially if the user sets `0` for the listening port; the listener returns the exact port.

File: www/jsonrpc/server.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	if s.server != nil {
+		s.server.Close()
+	}
+
+	if s.grpcClient != nil {
+		s.grpcClient.Close()
+	}
+}
+
+func (s *Server) Printf(format string, v ...interface{}) {
```
Comment (b00f) [inline www/jsonrpc/server.go:108]: I don't think this function is necessary anymore. In the previous implementation, it was used to print the internal logs of JSONRPC package.

File: www/grpc/gen/go/blockchain.pb.jgw.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+type BlockchainJsonRpcService struct {
+	client BlockchainClient
+}
+
+func NewBlockchainJsonRpcService(client BlockchainClient) BlockchainJsonRpcService {
+	return BlockchainJsonRpcService{
+		client: client,
+	}
+}
+
```
Comment (b00f) [inline www/grpc/gen/go/blockchain.pb.jgw.go:27]: You need to add a function here like:

```go
// RegisterBlockchainMethods registers the JSON-RPC methods for service Blockchain to "mux".
// The handlers forward requests to the grpc endpoint over "conn".
func RegisterTransactionHandler(ctx context.Context, mux *jgw.ServerMux, conn *grpc.ClientConn) error {
	return mux.RegisterMethods(ctx, BlockchainMethods(NewTransactionClient(conn)))
}

```

Comment (b00f) [reply www/grpc/gen/go/blockchain.pb.jgw.go:27]: You can give a type to: `func(ctx context.Context, params json.RawMessage) (interface{}, error)`

File: Makefile (hunk 1 above)
Comment (alidevjimmy) [reply Makefile:37]: I wanted to build daemon standalone and test it but there wasn't target for it.

File: config/config.go (hunk 2 above)
Comment (alidevjimmy) [reply config/config.go:125]: [3e1a4cad865e79547daba398fb9ad200cfbf07a2](https://github.com/pactus-project/pactus/pull/1183/commits/3e1a4cad865e79547daba398fb9ad200cfbf07a2)

File: www/grpc/buf/buf.gen.yaml (hunk 3 above)
Comment (alidevjimmy) [reply www/grpc/buf/buf.gen.yaml:37]: For now, it won't be required. I put it there for future plugin needs.

File: www/jsonrpc/server.go (hunk 4 above)
Comment (alidevjimmy) [reply www/jsonrpc/server.go]: https://github.com/pactus-project/pactus/pull/1183/commits/512d8dabf827c33e4f445b27ae9fd39f87beb40a

File: www/jsonrpc/server.go (hunk 5 above)
Comment (alidevjimmy) [reply www/jsonrpc/server.go:86]: How can i do that?
It seems http package only supports ListenAndServe methods.

File: www/jsonrpc/server.go (hunk 6 above)
Comment (alidevjimmy) [reply www/jsonrpc/server.go]: https://github.com/pactus-project/pactus/pull/1183/commits/1f0055a460a1bcdabfd3358ac802e61ef7ae779d

File: Makefile (hunk 1 above)
Comment (b00f) [reply Makefile:37]: Let's build all-in-on as i was before.

File: www/grpc/buf/buf.gen.yaml (hunk 3 above)
Comment (b00f) [reply www/grpc/buf/buf.gen.yaml:37]: So, let's remove it and add it when we need it

File: www/jsonrpc/server.go (hunk 5 above)
Comment (b00f) [reply www/jsonrpc/server.go]: As an example, look [here](https://github.com/pactus-project/pactus/blob/20d5d833394ea4f3b62cf14631b9584b4a9a1a28/www/http/server.go#L87)

File: go.mod [hunk 8]
```diff
@@ -1,12 +1,12 @@
 module github.com/pactus-project/pactus
 
-go 1.21
+go 1.22.0
```
Comment (b00f) [inline go.mod:3]: Please update go.mod in another PR.
We accept only atomic PRs

File: Makefile (hunk 1 above)
Comment (alidevjimmy) [reply Makefile:37]: sure

File: www/grpc/buf/buf.gen.yaml (hunk 3 above)
Comment (alidevjimmy) [reply www/grpc/buf/buf.gen.yaml:37]: Ok

File: node/node.go [hunk 9]
```diff
@@ -82,9 +84,9 @@ func NewNode(genDoc *genesis.Genesis, conf *config.Config,
 	if conf.GRPC.BasicAuthCredential != "" {
 		enableHTTPAuth = true
 	}
-
-	httpServer := http.NewServer(conf.HTTP, enableHTTPAuth)
 	grpcServer := grpc.NewServer(conf.GRPC, st, syn, net, consMgr)
+	httpServer := http.NewServer(conf.HTTP, enableHTTPAuth)
+	jsonrpcServer := jsonrpc.NewServer(conf.JSONRPC, enableHTTPAuth) // TODO: enable auth???
```
Comment (Ja7ad) [inline node/node.go:89]: Don't need to enable auth, because you need transcode json-rpc context to grpc metadata.

Please remove flag `enableHTTPAuth`.

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: Makefile [hunk 10]
```diff
@@ -74,7 +76,7 @@ check:
 # To avoid unintended conflicts with file names, always add to .PHONY
 # unless there is a reason not to.
 # https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html
-.PHONY: build build_gui
+.PHONY: build build_gui build_daemon build_wallet build_shell
```
Comment (b00f) [inline Makefile:79]: Please revert this line.

File: config/example_config.toml [hunk 11]
```diff
@@ -173,14 +173,22 @@
     # Default is `false`.
    ## enable_cors = false
 
+# `jsonrpc` configuration.
+[jsonrpc]
+  # `enable` indicates whether JSON-RPC service should be enabled or not.
+  # Default is `false`.
+ ## enable = false
+  # `listen` is the address to listen for incoming connections for JSON-RPC server.
```
Comment (b00f) [inline config/example_config.toml:181]: Add a new line here

**Review by b00f — APPROVED** [review]

File: node/node.go (hunk 9 above)
Comment (alidevjimmy) [reply node/node.go]: https://github.com/pactus-project/pactus/pull/1183/commits/a89973cbef09b33c059d9701b08897ecbfc08cda

File: config/example_config.toml (hunk 11 above)
Comment (alidevjimmy) [reply config/example_config.toml:182]: https://github.com/pactus-project/pactus/pull/1183/commits/75c3a0648c2499c8082d553aa70178c7b0658ae3

File: Makefile (hunk 10 above)
Comment (alidevjimmy) [reply Makefile]: https://github.com/pactus-project/pactus/pull/1183/commits/882407bf5b7ad0673475acb8ee1e634640a252d4

**Review by b00f — APPROVED** [review]

**Review by themantre — APPROVED** [review]

**Pull request merged**